### PR TITLE
feat(DDS): add resource dds lts log

### PR DESCRIPTION
**Pull request by 346010762**
**What this PR does / why we need it**:
add resource dds lts log

**Special notes for your reviewer**:

**Release note**:

```release-note
add resource dds lts log
```

## PR Checklist

* [ ] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST=./huaweicloud/services/acceptance/dds/ TESTARGS='-run TestAccDdsLtsLog_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dds/ -v -run TestAccDdsLtsLog_basic -timeout 360m -parallel 4
=== RUN   TestAccDdsLtsLog_basic
=== PAUSE TestAccDdsLtsLog_basic
=== CONT  TestAccDdsLtsLog_basic
--- PASS: TestAccDdsLtsLog_basic (927.98s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dds       928.021s

```

### CODE REVIEW

File: docs/resources/dds_lts_log.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+---
+subcategory: "Document Database Service (DDS)"
+---
+
+# huaweicloud_dds_lts_log
+
+Manages DDS LTS log resource within HuaweiCloud.
```
Comment (Jason-Zhang9309) [inline docs/resources/dds_lts_log.md]:
```suggestion
Manages a DDS LTS log resource within HuaweiCloud.
```

File: docs/resources/dds_lts_log.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource.
+  If omitted, the provider-level region will be used. Changing this creates a new resource.
+
+* `instance_id` - (Required, String, ForceNew) Specifies the ID of the DDS instance.
+  Changing this creates a new resource.
+
+* `log_type` - (Required, String, ForceNew) Specifies the type of the LTS log, e.g. **audit_log**.
```
Comment (Jason-Zhang9309) [inline docs/resources/dds_lts_log.md]:
```suggestion
* `log_type` - (Required, String, ForceNew) Specifies the type of the LTS log. The value can be **audit_log**.
```

File: docs/resources/dds_lts_log.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  Changing this creates a new resource.
+
+* `lts_stream_id` - (Required, String, ForceNew) Specifies the ID of the LTS log stream.
+  Changing this creates a new resource.
+
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID.
```
Comment (Jason-Zhang9309) [inline docs/resources/dds_lts_log.md]:
```suggestion
* `id` - The resource ID. The value is DDS instance ID.
```

File: docs/resources/dds_lts_log.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+* `create` - Default is 60 minutes.
+* `delete` - Default is 60 minutes.
+
+## Import
+
+The DDS LTS log config can be imported using `id`, e.g.
+
+```bash
+$ terraform import huaweicloud_dds_lts_log.test <id>
+```
```
Comment (Jason-Zhang9309) [inline docs/resources/dds_lts_log.md]:
```suggestion
The DDS LTS log config can be imported using instance ID, e.g.

```bash
$ terraform import huaweicloud_dds_lts_log.test <instance_id>
```
```

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.Errorf("error waiting for unassociating DDS (%s) with LTS log to become ready: %s ", d.Id(), err)
+	}
+
+	return resourceDdsLtsLogRead(ctx, d, meta)
+}
+
+func filterInstanceLtsConfig(instanceID string, resp interface{}) interface{} {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go]: you can rename this func to `FilterInstanceLtsConfig` and use it in the test

File: docs/resources/dds_lts_log.md (hunk 1 above)
Comment (346010762) [reply docs/resources/dds_lts_log.md]: done

File: docs/resources/dds_lts_log.md (hunk 2 above)
Comment (346010762) [reply docs/resources/dds_lts_log.md]: done

File: docs/resources/dds_lts_log.md (hunk 3 above)
Comment (346010762) [reply docs/resources/dds_lts_log.md]: done

File: docs/resources/dds_lts_log.md (hunk 4 above)
Comment (346010762) [reply docs/resources/dds_lts_log.md]: done

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go (hunk 5 above)
Comment (346010762) [reply huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go]: done

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+	err = json.Unmarshal(getDdsLtsLogRespJson, &getDdsLtsLogRespBody)
+	if err != nil {
+		return diag.Errorf("error unmarshaling DDS LTS configs: %s", err)
+	}
+
+	instanceLtsConfig := FilterInstanceLtsConfig(d.Id(), getDdsLtsLogRespBody)
+	instance := utils.PathSearch("instance", instanceLtsConfig, nil)
+	ltsConfigs := utils.PathSearch("lts_configs", instanceLtsConfig, make([]interface{}, 0)).([]interface{})
+	if len(ltsConfigs) == 0 || !utils.PathSearch("enabled", ltsConfigs[0], false).(bool) {
+		return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{}, "")
```
Comment (ShiChangkuo) [inline huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go:194]: if `enabled` is false, then the resource will be deleted. so is the `enabled` field meaningful?

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+	var getDdsLtsLogRespBody interface{}
+	err = json.Unmarshal(getDdsLtsLogRespJson, &getDdsLtsLogRespBody)
+	if err != nil {
+		return diag.Errorf("error unmarshaling DDS LTS configs: %s", err)
+	}
+
+	instanceLtsConfig := FilterInstanceLtsConfig(d.Id(), getDdsLtsLogRespBody)
+	instance := utils.PathSearch("instance", instanceLtsConfig, nil)
+	ltsConfigs := utils.PathSearch("lts_configs", instanceLtsConfig, make([]interface{}, 0)).([]interface{})
```
Comment (ShiChangkuo) [inline huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go]: suggest to use the following search expression to get LTS config object
```suggestion
	expression := fmt.Sprintf("lts_configs[?id=='%s'] | [0]", d.Id())
	ltsConfigs := utils.PathSearch(expression, getDdsLtsLogRespBody, nil)
```

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	var (
+		deleteDdsLtsLogHttpUrl = "v3/{project_id}/instances/logs/lts-configs"
+		deleteDdsLtsLogProduct = "dds"
+	)
+	deleteDdsLtsLogClient, err := cfg.NewServiceClient(deleteDdsLtsLogProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating DDS Client: %s", err)
```
Comment (ShiChangkuo) [inline huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go]: `client` should be in lowercase, please fix others
```suggestion
		return diag.Errorf("error creating DDS client: %s", err)
```

File: huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	var getDdsLtsLogRespBody interface{}
+	err = json.Unmarshal(getDdsLtsLogRespJson, &getDdsLtsLogRespBody)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshaling DDS LTS configs: %s", err)
+	}
+
+	instanceLtsConfig := dds.FilterInstanceLtsConfig(state.Primary.ID, getDdsLtsLogRespBody)
+	ltsConfigs := utils.PathSearch("lts_configs", instanceLtsConfig, make([]interface{}, 0)).([]interface{})
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go]:
```suggestion
	ltsConfigs ：= utils.PathSearch(fmt.Sprintf("instance_lts_configs[?instance.id=='%s']|[0].lts_configs|[0]"), getDdsLtsLogRespBody,nil)
```
and fix other places

File: huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDdsLtsLog_basic(rName),
+				Check: resource.ComposeTestCheckFunc(
+					rc.CheckResourceExists(),
+					resource.TestCheckResourceAttrPair(resourceName, "instance_id", "huaweicloud_dds_instance.instance", "id"),
+					resource.TestCheckResourceAttrPair(resourceName, "lts_group_id", "huaweicloud_lts_group.test", "id"),
+					resource.TestCheckResourceAttrPair(resourceName, "lts_stream_id", "huaweicloud_lts_stream.test", "id"),
+					resource.TestCheckResourceAttr(resourceName, "log_type", "audit_log"),
+					resource.TestCheckResourceAttrSet(resourceName, "enabled"),
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go]: you will wait for the config complete, so it is more better to check whether the value of enabled is true

File: huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  group_id    = huaweicloud_lts_group.test.id
+  stream_name = "%[2]s"
+}
+
+resource "huaweicloud_dds_lts_log" "test" {
+  instance_id   = huaweicloud_dds_instance.instance.id
+  log_type      = "audit_log"
+  lts_group_id  = huaweicloud_lts_group.test.id
+  lts_stream_id = huaweicloud_lts_stream.test.id
+
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go]: delete the empty line

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+		Target:     []string{"true"},
+		Refresh:    ddsLtsConfigRefreshFunc(createDdsLtsLogClient, instanceID),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      1 * time.Second,
+		MinTimeout: 2 * time.Second,
+	}
+
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.Errorf("error waiting for associating DDS (%s) with LTS log to become ready: %s ", instanceID, err)
```
Comment (houpeng80) [inline huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go]:
```suggestion
		return diag.Errorf("error waiting for associating DDS (%s) with LTS log to complete: %s ", instanceID, err)
```

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go (hunk 6 above)
Comment (346010762) [reply huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go:194]: removed the field enabled.

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go (hunk 7 above)
Comment (346010762) [reply huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go]: done

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go (hunk 8 above)
Comment (346010762) [reply huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go]: done

File: huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go (hunk 9 above)
Comment (346010762) [reply huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go]: done

File: huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go (hunk 10 above)
Comment (346010762) [reply huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go]: as we discussed, enabled field is deleted.

File: huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go (hunk 11 above)
Comment (346010762) [reply huaweicloud/services/acceptance/dds/resource_huaweicloud_dds_lts_log_test.go]: done

File: huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go (hunk 12 above)
Comment (346010762) [reply huaweicloud/services/dds/resource_huaweicloud_dds_lts_log.go]: done

**Comment by houpeng80** [comment]
/ lgtm
/approve

**Pull request merged**